concordium-cis2/src/lib.rs

```diff
@@ -1250,6 +1250,48 @@ pub struct OnReceivingCis2Params<T, A> {
     pub data:     AdditionalData,
 }
 
+/// Specific parameter type for a contract function which receives CIS2 tokens
+/// with a specific type D for the AdditionalData.
+// Note: For the serialization to be derived according to the CIS2
+// specification, the order of the fields cannot be changed.
+#[derive(Debug, SchemaType)]
+pub struct OnReceivingCis2DataParams<T, A, D> {
+    /// The ID of the token received.
+    pub token_id: T,
+    /// The amount of tokens received.
+    pub amount:   A,
+    /// The previous owner of the tokens.
+    pub from:     Address,
+    /// Some extra information which where sent as part of the transfer.
+    pub data:     D,
+}
+
+/// Deserial trait for OnReceivingCis2DataParams<T, A, D>.
+impl<T: Deserial, A: Deserial, D: Deserial> Deserial for OnReceivingCis2DataParams<T, A, D> {
+    fn deserial<R: Read>(source: &mut R) -> ParseResult<Self> {
+        let params: OnReceivingCis2Params<T, A> = source.get()?;
+        let additional_data_type: D = from_bytes(params.data.as_ref())?;
+        Ok(OnReceivingCis2DataParams {
+            token_id: params.token_id,
+            amount:   params.amount,
+            from:     params.from,
+            data:     additional_data_type,
+        })
+    }
+}
+
+/// Serial trait for OnReceivingCis2DataParams<T, A, D>.
+impl<T: Serial, A: Serial, D: Serial> Serial for OnReceivingCis2DataParams<T, A, D> {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        self.token_id.serial(out)?;
+        self.amount.serial(out)?;
+        self.from.serial(out)?;
+        (to_bytes(&self.data).len() as u16).serial(out)?;
+        out.write_all(&to_bytes(&self.data))?;
```

Comment:
```suggestion
        let add = AdditionalData(to_bytes(&self.data));
        add.serial(out)
```
avoid serializing the value twice please.